### PR TITLE
VizTooltips: Heatmap fixes and improvements

### PR DESCRIPTION
**Pull request by leeoniya**
fixes:

- timeseries tooltip was not showing `enum` type fields
- fixes numeric tooltip sorting (for real this time)
- heatmap tooltip was only activating within 30px proximity to y-center of each cell
- heatmap data links code was slow and :exploding_head:, and did not support field labels (this picks up where https://github.com/grafana/grafana/pull/83654 left off)

this refactors code related to:

- https://github.com/grafana/grafana/pull/60039
- https://github.com/grafana/grafana/pull/71016

### CODE REVIEW

File: public/app/plugins/panel/timeseries/TimeSeriesTooltip.tsx [hunk 1]
```diff
@@ -56,7 +56,7 @@ export const TimeSeriesTooltip = ({
     seriesIdx,
     mode,
     sortOrder,
-    (field) => field.type === FieldType.number
+    (field) => field.type === FieldType.number || field.type === FieldType.enum
```
Comment (leeoniya) [inline public/app/plugins/panel/timeseries/TimeSeriesTooltip.tsx:59]: enum field fix

File: public/app/plugins/panel/heatmap/utils.ts [hunk 2]
```diff
@@ -585,6 +586,10 @@ export function prepConfig(opts: PrepConfigOpts) {
 
       return hRect && seriesIdx === hRect.sidx ? hRect.didx : null;
     },
+    focus: {
+      prox: 1e3,
+      dist: (u, seriesIdx) => (hRect?.sidx === seriesIdx ? 0 : Infinity),
+    },
```
Comment (leeoniya) [inline public/app/plugins/panel/heatmap/utils.ts:592]: heatmap cell hover prox fix

File: packages/grafana-ui/src/components/VizTooltip/utils.ts [hunk 3]
```diff
@@ -85,7 +85,7 @@ export const getContentItems = (
 ): VizTooltipItem[] => {
   let rows: VizTooltipItem[] = [];
 
-  let allNumeric = false;
+  let allNumeric = true;
```
Comment (leeoniya) [inline packages/grafana-ui/src/components/VizTooltip/utils.ts:88]: sorting fix

File: public/app/plugins/panel/heatmap/utils.ts [hunk 4]
```diff
@@ -772,6 +777,8 @@ export function heatmapPathsPoints(opts: PointsBuilderOpts, exemplarColor: strin
         let fillPaths = [points];
         let fillPalette = [exemplarColor ?? 'rgba(255,0,255,0.7)'];
 
+        let yShift = yLayout === HeatmapCellLayout.le ? -0.5 : yLayout === HeatmapCellLayout.ge ? 0.5 : 0;
```
Comment (leeoniya) [inline public/app/plugins/panel/heatmap/utils.ts:780]: this fixes vertical exemplar positioning in each cell. previously we assumed `le` (cell down from tick), but with ordinal it is centered, and can be manually set as well via panel opts.

File: public/app/plugins/panel/heatmap/utils.ts [hunk 5]
```diff
@@ -782,10 +789,7 @@ export function heatmapPathsPoints(opts: PointsBuilderOpts, exemplarColor: strin
           let isSparseHeatmap = scaleY.distr === 3 && scaleY.log === 2;
 
           if (!isSparseHeatmap) {
-            yVal -= 0.5; // center vertically in bucket (when tiles are le)
-            // y-randomize vertically to distribute exemplars in same bucket at same time
-            let randSign = Math.round(Math.random()) * 2 - 1;
-            yVal += randSign * 0.5 * Math.random();
```
Comment (leeoniya) [inline public/app/plugins/panel/heatmap/utils.ts:788]: this removes exemplar randomization within each cell and provides stability, but also maybe overlap. probably a better trade-off. we can look at clustering or multi-hover in the future.

**Review by baldm0mma — COMMENTED** [review]
- [x] timeseries tooltip was not showing enum type fields
- [x] fixes numeric tooltip sorting (for real this time)
- [x] heatmap tooltip was only activating within 30px proximity to y-center of each cell
- [x] heatmap data links code was slow and did not support field labels
- [-] tooltips work as expected when `newVizTooltips` FF is toggled on and off - for this I did notice one thing: when hovering a value with the **_old_** tooltips, the color scale at the bottom would display the hovered value on the range, but with the new tooltips, this functionality seems to be missing. Is this expected?
Screenshot 2024-03-06 at 2 39 25 PM
Screenshot 2024-03-06 at 2 38 42 PM

**Comment by leeoniya** [comment]
> tooltips work as expected when `newVizTooltips` FF is toggled on and off - for this I did notice one thing: when hovering a value with the **_old_** tooltips, the color scale at the bottom would display the hovered value on the range, but with the new tooltips, this functionality seems to be missing. Is this expected?

yes, this is expected for now. this was the only panel with some kind of live legend-tooltip behavior, so was too much of a snowflake to keep in right now. we will revisit this concept holistically for all panels as part of the legends work.

the live color scale was moved into the tooltip as a toggle-able option in panel edit (same as live histogram in tooltip).

**Review by adela-almasan — APPROVED** [review]

**Pull request merged**